incubator/hnc/internal/reconcilers/hnc_config.go

```diff
@@ -98,6 +110,92 @@ func (r *ConfigReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	return ctrl.Result{}, syncErr
 }
 
+// reconcileTypes makes sure there's no dup and the types exist. Update the type
+// set with GR to GVK mappings.
+func (r *ConfigReconciler) reconcileTypes(inst *api.HNCConfiguration) error {
+	// Get all resources for all groups.
+	allRes, err := r.getAllResources()
+	if err != nil {
+		r.Log.Error(err, "while trying to get all resources")
+		return err
+	}
+	// exist keeps track of existing types in the `config` singleton.
+	exist := grSet{}
+	r.activeGRs = grSet{}
+	r.activeGVKs = gvkSet{}
+	for _, t := range inst.Spec.Types {
+		// If there are multiple configurations of the same type, we will follow the first
+		// configuration and ignore the rest.
+		if !r.ensureNoDuplicateTypeConfigurations(inst, t, exist) {
+			continue
+		}
+		// Look if the resource exists in the API server.
+		gr := schema.GroupResource{Group: t.Group, Resource: t.Resource}
+		if err := r.addType(gr, allRes); err != nil {
+			// If the type is not found, log error and write conditions but don't
+			// early exit since the other types can still be reconciled.
+			r.Log.Error(err, "while trying to get the resource")
+			r.writeCondition(inst, api.TypeNotFound, err.Error())
+		}
+	}
+	return nil
+}
+
+func (r *ConfigReconciler) getAllResources() ([]*restmapper.APIGroupResources, error) {
+	// Get all resources for all groups.
+	dc, err := discovery.NewDiscoveryClientForConfig(r.Manager.GetConfig())
+	if err != nil {
+		r.Log.Error(err, "while creating discovery client")
+		return nil, err
+	}
+	return restmapper.GetAPIGroupResources(dc)
+}
+
+// ensureNoDuplicateTypeConfigurations checks whether the configuration of a
+// type already exists and sets a condition in the status if the type exists.
+// The method returns true if the type configuration does not exist; otherwise,
+// returns false.
+func (r *ConfigReconciler) ensureNoDuplicateTypeConfigurations(inst *api.HNCConfiguration, t api.TypeSynchronizationSpec,
+	mp grSet) bool {
+	gr := schema.GroupResource{Group: t.Group, Resource: t.Resource}
+	if _, ok := mp[gr]; !ok {
+		mp[gr] = schema.GroupVersionKind{}
+		return true
+	}
+	specMsg := fmt.Sprintf("Group: %s, Resource: %s, Mode: %s", t.Group, t.Resource, t.Mode)
+	r.Log.Info(fmt.Sprintf("Ignoring the configuration: %s", specMsg))
+	msg := fmt.Sprintf("Ignore the configuration: %s because the configuration of the type already exists; "+
+		"only the first configuration will be applied", specMsg)
+	r.writeCondition(inst, api.MultipleConfigurationsForOneType, msg)
+	return false
+}
+
+// addType sets the type with GR to GVK mapping. Return error if the resource is
+// not found.
+func (r *ConfigReconciler) addType(gr schema.GroupResource, allRes []*restmapper.APIGroupResources) error {
+	// Look for a matching resource from all resources.
+	for _, groupedResources := range allRes {
```

Comment:
This code looks duplicated in the validator. Factor this out so that it's exactly the same code running in both places?